<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Ciangene by CianMurphy</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">Ciangene</h1>
        <p class="header"></p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/CianMurphy/ciangene/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/CianMurphy/ciangene/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/CianMurphy/ciangene">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/CianMurphy">CianMurphy</a></p>


      </header>
      <section>
        <h1>
<a id="ciangene" class="anchor" href="#ciangene" aria-hidden="true"><span class="octicon octicon-link"></span></a>ciangene</h1>

<p><a href="https://gitter.im/CianMurphy/ciangene?utm_source=badge&amp;utm_medium=badge&amp;utm_campaign=pr-badge&amp;utm_content=badge"><img src="https://badges.gitter.im/Join%20Chat.svg" alt="Join the chat at https://gitter.im/CianMurphy/ciangene"></a>
Order of scripts: in progress</p>

<ul>
<li>
<p>first.step.R</p>

<pre><code>Takes the snpStats files from the data release and turn into a numeric matrix
Same thing but with depth, prepares the kinshpi matrix computation
Takes the list of clean variants and outputs the filtered result
</code></pre>
</li>
<li>
<p>variant_filtering</p>

<pre><code>variant_filtering/snp_subet_extCtrls.R
</code></pre>
</li>
<li>
<p>convert_missingNonmissing_matrix_to_plink.sh</p>

<pre><code>turns NA/non NA into 0-1 for LDAK/plink use
</code></pre>
</li>
<li>
<p>make_kinships_new.sh</p>

<pre><code>###Computes variant weights, then computes kinship matrix based on depth and missing/non-missing
###Three sets of weights: two kinship matrices and genotypes.
###Why? Need to understand why we use weights from genotype matrix.
</code></pre>
</li>
<li>
<p>new_make_phenotype_file.R</p>

<pre><code>Creates a large matrix of phenotypes, grouped by names basically.
</code></pre>
</li>
<li>
<p>check_tk_residuals.sh</p>

<pre><code>Checks how much variance TK explains for all phenos (A). Creates new pheno files, using residuals from (A) as phenos. 
</code></pre>
</li>
<li>
<p>LDAK/run_ldak_on_all_phenos.sh</p>

<pre><code>gene tests - all UCLex phenos, with and without kinship
</code></pre>
</li>
<li>
<p>plink_single_variant_tests.sh</p>

<pre><code>single variant tests - Fisher, logistic regression with and without techPC covariates 
</code></pre>
</li>
<li>
<p>run_fastLMM_on_all_phenotypes.sh </p>

<pre><code>first, run prepare_kinship_matrix_for_fastLMM.R

single variant tests - FastLMM - Linear Regression / Mixed model with and without techKin
</code></pre>
</li>
<li>
<p>merge_gene_tests.sh</p>

<pre><code>Doesn't just merge. Also uses permuted pvalue in regress1..regressN to define null distribution (permuted pheno with kin) to calc final pvalue in regressALL. Script hangs when run on just one chunk/gene (eg as per power studies). 
</code></pre>
</li>
<li><p>plot_singleVariant_results.R</p></li>
</ul>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		
  </body>
</html>
